_posts/2015-05-11-traits.md

```diff
@@ -0,0 +1,420 @@
+abstracting out a common interface satisfied by more than one type.
+
+### Static dispatch
+
+Things get more interesting on the other side -- consuming a trait. The most
+common way of doing so is through *generics*:
+
+```rust
+fn print_hash<T: Hash>(t: &T) {
+    println!("The hash is {}", t.hash())
+}
+```
+
+The `print_hash` function is generic over an unknown type `T`, but requires that
+`T` implements the `Hash` trait. That means we can use it with `bool` and `i64`
+values:
+
+```rust
+print_hash(&true);      // instantiates T = bool
+print_hash(&12_i64);    // instantiates T = i64
+```
+
+**Generics are compiled away, resulting in static dispatch**. That is, as with
+C++ templates, the compiler will generate *two copies* of the `print_hash`
+method to handle the above code, one for each concrete argument type.  That in
+turn means that the internal call to `t.hash()` -- the point where the
+abstraction is actually used -- has zero cost: it will be compiled to a direct,
+static call to the relevant implementation.
+
+This compilation model isn't so useful for a function like `print_hash`, but
+it's *very* useful for more realistic uses of hashing. Suppose we also introduce
+a trait for equality comparison:
+
+```rust
+trait Eq {
+    fn eq(&self, other: &Self) -> bool;
+}
+```
+
+(The reference to `Self` here will resolve to whatever type we implement the
+trait for; in `impl Eq for bool` it will refer to `bool`.)
+
+We can then define a hash map that is generic over a type `T` implementing both
+`Hash` and `Eq`:
+
+```rust
+struct HashMap<Key: Hash + Eq, Value> { ... }
+```
+
+The static compilation model for generics will then yield several benefits:
+
+* Each use of `HashMap` with concrete `Key` and `Value` types will result in a
+  different concrete `HashMap` type, which means that `HashMap` can lay out the
+  keys and values in-line (without indirection) in its buckets. This saves on
+  space and indirections, and improves cache locality.
+
+* Each method on `HashMap` will likewise generate specialized code. That means
+  there is no extra cost dispatching to calls to `hash` and `eq`, as above.  It
+  also means that the optimizer gets to work with the fully concrete code --
+  that is, from the point of view of the optimizer, *there is no abstraction*.
+
+Altogether, just as in C++ templates, these aspects of generics mean that you
+can write quite high-level abstractions that are *guaranteed* to compile down to
+fully concrete code that "you couldn't hand code any better".
+
+**But, unlike with C++ templates, clients of traits are fully type-checked in
+advance**.  That is, when you compile `HashMap` in isolation, its code is
+checked *once* for type correctness against the abstract `Hash` and `Eq` traits,
+rather than being checked repeatedly when applied to concrete types. That means
+earlier, clearer compilation errors for library authors, and less typechecking
+overhead for clients.
```

Comment:
It may be worth explicitly spelling out here that "less typechecking overhead" can directly lead to faster compile times (in theory)